File: data/derived_cot/rq1_traces/translation/r1/py_to_cpp/code_output/PushBoxGame.cpp
#include <vector>
#include <utility>
#include <algorithm>

class PushBoxGame {
public:
    std::vector<std::string> map_;
    int player_row_;
    int player_col_;
    std::vector<std::pair<int, int>> targets_;
    std::vector<std::pair<int, int>> boxes_;
    int target_count_;
    bool is_game_over_;

    PushBoxGame(const std::vector<std::string>& map) : 
        map_(map), player_row_(0), player_col_(0), target_count_(0), is_game_over_(false) {
        initGame();
    }

private:
    void initGame() {
        for (int row = 0; row < map_.size(); row++) {
            for (int col = 0; col < map_[row].size(); col++) {
                char c = map_[row][col];
                if (c == 'O') {
                    player_row_ = row;
                    player_col_ = col;
                } else if (c == 'G') {
                    targets_.push_back(std::make_pair(row, col));
                    target_count_++;
                } else if (c == 'X') {
                    boxes_.push_back(std::make_pair(row, col));
                }
            }
        }
    }

public:
    bool checkWin() {
        int boxOnTargetCount = 0;
        for (const auto& box : boxes_) {
            if (std::find(targets_.begin(), targets_.end(), box) != targets_.end()) {
                boxOnTargetCount++;
            }
        }
        if (boxOnTargetCount == target_count_) {
            is_game_over_ = true;
            return true;
        }
        return false;
    }

    bool move(char direction) {
        int new_player_row = player_row_;
        int new_player_col = player_col_;

        switch (direction) {
            case 'w': new_player_row--; break;
            case 's': new_player_row++; break;
            case 'a': new_player_col--; break;
            case 'd': new_player_col++; break;
            default: 
                break;
        }

        if (new_player_row < 0 || new_player_row >= map_.size()) {
            return checkWin();
        }
        if (new_player_col < 0 || new_player_col >= map_[new_player_row].size()) {
            return checkWin();
        }

        if (map_[new_player_row][new_player_col] == '#') {
            return checkWin();
        }

        auto it = std::find(boxes_.begin(), boxes_.end(), std::make_pair(new_player_row, new_player_col));
        if (it != boxes_.end()) {
            int new_box_row = new_player_row + (new_player_row - player_row_);
            int new_box_col = new_player_col + (new_player_col - player_col_);

            if (new_box_row < 0 || new_box_row >= map_.size()) {
                return checkWin();
            }
            if (new_box_col < 0 || new_box_col >= map_[new_box_row].size()) {
                return checkWin();
            }

            if (map_[new_box_row][new_box_col] == '#') {
                return checkWin();
            }

            boxes_.erase(it);
            boxes_.push_back(std::make_pair(new_box_row, new_box_col));

            player_row_ = new_player_row;
            player_col_ = new_player_col;
        } else {
            player_row_ = new_player_row;
            player_col_ = new_player_col;
        }

        return checkWin();
    }
};

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/cpp_to_py/PushBoxGame.py
class PushBoxGame:
    def __init__(self, map=None):
        if map is None:
            self.map = []
        else:
            self.map = map[:]  # Make a shallow copy of the list
        self.player_row = 0
        self.player_col = 0
        self.targets = []      # List of tuples (row, col)
        self.boxes = []        # List of tuples (row, col)
        self.target_count = 0
        self._is_game_over = False
        self.init_game()
    
    def gat_map(self):
        return self.map[:]  # Return a copy of the map
    
    def is_game_over(self):
        return self._is_game_over
    
    def get_player_col(self):
        return self.player_col
    
    def get_player_row(self):
        return self.player_row
    
    def get_targets(self):
        return self.targets[:]  # Return a copy
    
    def get_boxes(self):
        return self.boxes[:]  # Return a copy
    
    def get_target_count(self):
        return self.target_count
    
    def init_game(self):
        for row_idx, row in enumerate(self.map):
            for col_idx, cell in enumerate(row):
                if cell == 'O':
                    self.player_row = row_idx
                    self.player_col = col_idx
                elif cell == 'G':
                    self.targets.append((row_idx, col_idx))
                    self.target_count += 1
                elif cell == 'X':
                    self.boxes.append((row_idx, col_idx))
    
    def check_win(self):
        box_on_target_count = 0
        for box in self.boxes:
            if box in self.targets:
                box_on_target_count += 1
        if box_on_target_count == self.target_count:
            self._is_game_over = True
        return self._is_game_over
    
    def move(self, direction):
        dr = 0
        dc = 0
        if direction == 'w':
            dr = -1
        elif direction == 's':
            dr = 1
        elif direction == 'a':
            dc = -1
        elif direction == 'd':
            dc = 1
        
        new_player_row = self.player_row + dr
        new_player_col = self.player_col + dc
        
        if not (0 <= new_player_row < len(self.map) and 0 <= new_player_col < len(self.map[0])):
            return self.check_win()
        
        if self.map[new_player_row][new_player_col] == '#':
            return self.check_win()
        
        new_pos = (new_player_row, new_player_col)
        if new_pos in self.boxes:
            new_box_row = new_player_row + dr
            new_box_col = new_player_col + dc
            if not (0 <= new_box_row < len(self.map) and 0 <= new_box_col < len(self.map[0])):
                return self.check_win()
            if self.map[new_box_row][new_box_col] == '#' or (new_box_row, new_box_col) in self.boxes:
                return self.check_win()
            
            self.boxes.remove(new_pos)
            self.boxes.append((new_box_row, new_box_col))
            self.player_row = new_player_row
            self.player_col = new_player_col
        else:
            self.player_row = new_player_row
            self.player_col = new_player_col
        
        return self.check_win()